larq_compute_engine/mlir/transforms/optimize.cc

```diff
@@ -68,123 +68,88 @@ DenseElementsAttr Bitpack(PatternRewriter& builder, Attribute x) {
 
 #include "larq_compute_engine/mlir/transforms/generated_optimize.inc"
 
-struct SetBconvReadWriteBitpacked : public OpRewritePattern<TF::LceBconv2dOp> {
-  using OpRewritePattern<TF::LceBconv2dOp>::OpRewritePattern;
-
-  LogicalResult matchAndRewrite(TF::LceBconv2dOp outer_bconv_op,
-                                PatternRewriter& rewriter) const override {
-    /*************
-     * Match ops *
-     *************/
-
-    auto intermediate_maxpool_op = dyn_cast_or_null<TFL::MaxPool2DOp>(
-        outer_bconv_op.input().getDefiningOp());
-
-    if (intermediate_maxpool_op &&
-        intermediate_maxpool_op.fused_activation_function() != "NONE") {
-      return failure();
-    }
-
-    auto inner_bconv_op =
-        intermediate_maxpool_op
-            ? dyn_cast_or_null<TF::LceBconv2dOp>(
-                  intermediate_maxpool_op.input().getDefiningOp())
-            : dyn_cast_or_null<TF::LceBconv2dOp>(
-                  outer_bconv_op.input().getDefiningOp());
-
-    if (inner_bconv_op && inner_bconv_op.padding() == "SAME" &&
-        inner_bconv_op.pad_values() != 1) {
-      // We don't return `failure()` here because we still want to match the
-      // pattern `binary maxpool -> binary conv`.
-      inner_bconv_op = nullptr;
-    }
-
-    if (!inner_bconv_op && !intermediate_maxpool_op) return failure();
-
-    if (!outer_bconv_op.input().hasOneUse() ||
-        (intermediate_maxpool_op && inner_bconv_op &&
-         !intermediate_maxpool_op.input().hasOneUse())) {
-      return failure();
-    }
+/**
+ * Returns `false` if some assumption does not hold (the consumer should then
+ * consider the rewrite pattern match to have failed).
+ */
+bool getBitpackedType(PatternRewriter& rewriter, ShapedType existing_type,
+                      RankedTensorType& bitpacked_type) {
```

Comment:
How do you feel about changing this to something like:
```cpp
#include "llvm/ADT/Optional.h"

llvm::Optional<RankedTensorType> (PatternRewriter& rewriter, ShapedType existing_type) {
```

I think this would make the checks easier to read and would not require us to initialize `RankedTensorType` before using this function.